content/article/creating-midi-music-with-perl.md

```diff
@@ -0,0 +1,443 @@
+
+```
+timidity -c ~/timidity.cfg some.mid -Ow -o - | ffmpeg -i - -acodec libmp3lame -ab 64k some.mp3
+```
+
+But wait! You can also generate and play MIDI in real-time with the [MIDI::RtMidi::FFI::Device]({{< mcpan "MIDI::RtMidi::FFI::Device" >}}) and [MIDI::RtMidi::ScorePlayer]({{< mcpan "MIDI::RtMidi::ScorePlayer" >}}) modules. :D
+
+Back to Creating Music!
+-----------------------
+
+So far, we have encountered the "Setup, Play, and Write" algorithm. Next we shall replace the "Play" bit with "Sync" and play the bass and treble parts simultaneously:
+
+```perl
+use MIDI::Util qw(setup_score);
+
+my $score = setup_score();
+
+$score->synch(
+  sub { bass($score) },
+  sub { treble($score) },
+);
+
+$score->write_score("$0.mid");
+
+sub bass {
+  my ($score) = @_;
+  for my $note (qw(C3 F3 G3 C4)) {
+    $score->n('qn', $note);
+  }
+}
+
+sub treble {
+  my ($score) = @_;
+  for my $note (qw(C4 D4 E4 F4)) {
+    $score->n('en', $note);
+    $score->r('en');
+  }
+}
+```
+
+This algorithm is not especially clever, but illustrates the basics.
+
+If we want to repeat the phrase, just add a `for` loop to the `synch`:
+
+```perl
+$score->synch(
+  sub { bass($score) },
+  sub { treble($score) },
+) for 1 .. 4;
+```
+
+{{< audio src="/images/creating-midi-music-with-perl/audio-2.mp3" type="audio/mpeg" >}}
+
+Setting Channels, Patches, Volume, and Tempo
+--------------------------------------------
+
+Use the `MIDI::Util::set_chan_patch()` function to set the channel and the patch.
+
+To set the tempo in "beats per minute", give a `bpm` to the `setup_score()` function. To set an individual note volume, add `"v$num"` as an argument to the `$score->n()` method, where `$num` is an integer from 0 to 127. Here is code that sets these parameters:
+
+```perl
+use MIDI::Util qw(setup_score set_chan_patch);
+
+my $bpm = shift || 120;
+
+my $score = setup_score(bpm => $bpm);
+
+$score->synch(\&bass, \&treble);
+
+$score->write_score("$0.mid");
+
+sub bass {
+  set_chan_patch($score, 0, 35);
+  for my $note (qw(C3 F3 G3 C4)) {
+    $score->n('qn', $note, 'v127');
+  }
+}
+
+sub treble {
+  set_chan_patch($score, 1, 0);
+  for my $note (qw(C4 D4 E4 F4)) {
+    $score->n('en', $note, 'v110');
+    $score->r('en');
+  }
+}
+```
+
+Notice that the `\&` syntax is used for the `synch`'d subs. This makes the bass and treble subroutines use the `$score` object as a global variable. Sometimes this is done... Some may hate this. Fortunately, it's optional.
+
+Selecting Pitches
+-----------------
+
+What if we want the program to choose notes at random, to add to the score? Here is a simple example:
+
+```perl
+sub treble {
+  set_chan_patch($score, 1, 0);
+
+  my @pitches = (60, 62, 64, 65, 67, 69, 71, 72);
+
+  for my $n (1 .. 4) {
+    my $pitch = $pitches[int rand @pitches];
+    $score->n('en', $pitch);
+    $score->r('en');
+  }
+}
+```
+
+For MIDI-Perl, the named note with octave `"C4"` and the MIDI number `"60"` are identical, as shown above.
```

Comment:
Maybe link to (or include) a chart which maps out these relationships?

Reply:
Done